src/runtime/set.cpp

```diff
@@ -100,51 +109,90 @@ BoxedSet* makeNewSet(BoxedClass* cls, Box* container) {
     BoxedSet* rtn = new (cls) BoxedSet();
 
     if (container) {
-        for (Box* e : container->pyElements()) {
-            rtn->s.insert(e);
+        if (PyAnySet_Check(container)) {
+            for (auto&& elt : ((BoxedSet*)container)->s) {
+                rtn->s.insert(elt);
+            }
+        } else if (PyDict_CheckExact(container)) {
+            for (auto&& elt : ((BoxedDict*)container)->d) {
+                rtn->s.insert(elt.first);
+            }
+        } else {
+            for (auto elt : container->pyElements()) {
+                rtn->s.insert(elt);
+            }
         }
     }
     return rtn;
 }
 
-Box* frozensetNew(Box* _cls, Box* container) {
+Box* frozensetNew(Box* _cls, Box* container, BoxedDict* kwargs) {
     RELEASE_ASSERT(_cls->cls == type_cls, "");
     BoxedClass* cls = static_cast<BoxedClass*>(_cls);
     RELEASE_ASSERT(isSubclass(cls, frozenset_cls), "");
 
     // Some optimizations from CPython: frozensets can be shared:
     if (_cls == frozenset_cls) {
-        if (!container) {
+        if (!_PyArg_NoKeywords("frozenset()", kwargs)) {
+            throwCAPIException();
+        }
+
+        BoxedSet* result = makeNewSet(cls, container);
```

Comment:
Can we use the CPython logic for this section?  They have this:
```C
    if (iterable != NULL) {
        /* frozenset(f) is idempotent */
        if (PyFrozenSet_CheckExact(iterable)) {
            Py_INCREF(iterable);
            return iterable;
        }
        result = make_new_set(type, iterable);
        if (result == NULL || PySet_GET_SIZE(result))
            return result;
        Py_DECREF(result);
    }
    /* The empty frozenset is a singleton */
    if (emptyfrozenset == NULL)
        emptyfrozenset = make_new_set(type, NULL);
    Py_XINCREF(emptyfrozenset);
    return emptyfrozenset;
```

It's pretty similar, but the different ordering means that they can skip allocating a frozenset instance if they can.  (the code here will call makeNewSet regardless.)